src/Nerdbank.Streams/NestedStream.cs

```diff
@@ -73,52 +97,85 @@ public override long Position
 
             if (count == 0)
             {
                 return 0;
             }
 
             int bytesRead = await this.underlyingStream.ReadAsync(buffer, offset, count).ConfigureAwaitRunInline();
-            this.length -= bytesRead;
+            this.remainingBytes -= bytesRead;
             return bytesRead;
         }
 
         /// <inheritdoc />
         public override int Read(byte[] buffer, int offset, int count)
         {
-            count = (int)Math.Min(count, this.length);
+            count = (int)Math.Min(count, this.remainingBytes);
 
             if (count == 0)
             {
                 return 0;
             }
 
             int bytesRead = this.underlyingStream.Read(buffer, offset, count);
-            this.length -= bytesRead;
+            this.remainingBytes -= bytesRead;
             return bytesRead;
         }
 
 #if SPAN_BUILTIN
         /// <inheritdoc />
         public override async ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken = default)
         {
-            buffer = buffer.Slice(0, (int)Math.Min(buffer.Length, this.length));
+            buffer = buffer.Slice(0, (int)Math.Min(buffer.Length, this.remainingBytes));
 
             if (buffer.IsEmpty)
             {
                 return 0;
             }
 
             int bytesRead = await this.underlyingStream.ReadAsync(buffer, cancellationToken).ConfigureAwait(false);
-            this.length -= bytesRead;
+            this.remainingBytes -= bytesRead;
             return bytesRead;
         }
 #endif
 
         /// <inheritdoc />
-        public override long Seek(long offset, SeekOrigin origin) => throw this.ThrowDisposedOr(new NotSupportedException());
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            Verify.NotDisposed(this);
+
+            if (!this.CanSeek)
+            {
+                throw new NotSupportedException("This stream does not support seeking.");
+            }
+
+            long boundedOffset = 0;
+
+            if (origin == SeekOrigin.Current)
+            {
+                boundedOffset = offset;
+            }
+            else if (origin == SeekOrigin.End)
+            {
+                boundedOffset = this.length + offset - this.Position;
+            }
+            else if (origin == SeekOrigin.Begin)
+            {
+                boundedOffset = offset - this.Position;
+            }
+
+            boundedOffset =
+                boundedOffset >= 0 ?
+                    Math.Min(boundedOffset, this.remainingBytes) :
+                    Math.Max(boundedOffset, -1 * this.Position);
+
+            long currentPosition = this.underlyingStream.Position;
+            long newPosition = this.underlyingStream.Seek(Math.Min(this.remainingBytes, boundedOffset), SeekOrigin.Current);
+            this.remainingBytes -= newPosition - currentPosition;
+            return this.Position;
```

Comment:
TBH I'm having trouble following the logic here. And since it needs to be altered to match docs and other Stream behavior anyway, can it be rewritten as a `switch` statement based on `origin` where we share less code across each branch (and thus fewer Math.Min/Max expressions nested inside conditionals)?